modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/ExpressionFactoryImpl.java

```diff
@@ -765,36 +795,44 @@ private class RangeConditionImpl implements RangeCondition<RowT> {
         /** Upper row. */
         private @Nullable RowT upperRow;
 
-        /** Cached skip range flag. */
-        private Boolean skip;
+        /** Lower bound row factory. */
+        private final RowFactory<RowT> lowerFactory;
 
-        /** Row factory. */
-        private final RowFactory<RowT> factory;
+        /** Upper bound row factory. */
+        private final RowFactory<RowT> upperFactory;
+
+        /** Cached skip range flag. */
+        private @Nullable Boolean skip;
 
         private RangeConditionImpl(
-                SingleScalar lowerBound,
-                SingleScalar upperBound,
+                List<RexNode> lower,
+                List<RexNode> upper,
                 boolean lowerInclude,
                 boolean upperInclude,
-                RowFactory<RowT> factory
+                RelDataType rowTypeLower,
+                RelDataType rowTypeUpper,
+                RowFactory<RowT> lowerFactory,
+                RowFactory<RowT> upperFactory
         ) {
-            this.lowerBound = lowerBound;
-            this.upperBound = upperBound;
+            this.lowerBound = scalar(lower, rowTypeLower);
+            this.upperBound = scalar(upper, rowTypeUpper);
             this.lowerInclude = lowerInclude;
             this.upperInclude = upperInclude;
-            this.factory = factory;
+
+            this.lowerFactory = lowerFactory;
+            this.upperFactory = upperFactory;
         }
 
         /** {@inheritDoc} */
         @Override
-        public RowT lower() {
-            return lowerRow != null ? lowerRow : (lowerRow = getRow(lowerBound));
+        public @Nullable RowT lower() {
+            return lowerRow != null ? lowerRow : getRow(lowerBound, lowerFactory);
```

Comment:
As I mentioned earlier, with this change, lowerRow/upperRow is always null now.

Reply:
got it, thnks !